Content.Shared/Xenoarchaeology/Artifact/SharedXenoArtifactSystem.Node.cs

```diff
@@ -0,0 +1,314 @@
+
+public abstract partial class SharedXenoArtifactSystem
+{
+    private EntityQuery<XenoArtifactComponent> _xenoArtifactQuery;
+    private EntityQuery<XenoArtifactNodeComponent> _nodeQuery;
+
+    private void InitializeNode()
+    {
+        SubscribeLocalEvent<XenoArtifactNodeComponent, MapInitEvent>(OnNodeMapInit);
+
+        _xenoArtifactQuery = GetEntityQuery<XenoArtifactComponent>();
+        _nodeQuery = GetEntityQuery<XenoArtifactNodeComponent>();
+    }
+
+    private void OnNodeMapInit(Entity<XenoArtifactNodeComponent> ent, ref MapInitEvent args)
+    {
+        SetNodeDurability((ent, ent), ent.Comp.MaxDurability - ent.Comp.InitialDurabilityVariation.Next(RobustRandom));
+    }
+
+    public XenoArtifactNodeComponent XenoArtifactNode(EntityUid uid)
+    {
+        return _nodeQuery.Get(uid);
+    }
+
+    public void SetNodeUnlocked(Entity<XenoArtifactNodeComponent?> ent)
+    {
+        if (!Resolve(ent, ref ent.Comp))
+            return;
+
+        if (ent.Comp.Attached is not { } netArtifact)
+            return;
+        var artifact = GetEntity(netArtifact);
+
+        if (!TryComp<XenoArtifactComponent>(artifact, out var artifactComponent))
+            return;
+
+        SetNodeUnlocked((artifact, artifactComponent), (ent, ent.Comp));
+    }
+
+    public void SetNodeUnlocked(Entity<XenoArtifactComponent> artifact, Entity<XenoArtifactNodeComponent> node)
+    {
+        if (!node.Comp.Locked)
+            return;
+
+        node.Comp.Locked = false;
+        RebuildCachedActiveNodes((artifact, artifact));
+        Dirty(node);
+    }
+
+    /// <summary>
+    /// Adds to the nodes durability by the specified value.
+    /// </summary>
+    /// <param name="ent"></param>
+    /// <param name="delta"></param>
+    public void AdjustNodeDurability(Entity<XenoArtifactNodeComponent?> ent, int delta)
+    {
+        if (!Resolve(ent, ref ent.Comp))
+            return;
+        SetNodeDurability(ent, ent.Comp.Durability + delta);
+    }
+
+    /// <summary>
+    /// Sets a node's durability to the specified value.
+    /// </summary>
+    public void SetNodeDurability(Entity<XenoArtifactNodeComponent?> ent, int durability)
+    {
+        if (!Resolve(ent, ref ent.Comp))
+            return;
+        ent.Comp.Durability = Math.Clamp(durability, 0, ent.Comp.MaxDurability);
+        UpdateNodeResearchValue((ent, ent.Comp));
+        Dirty(ent);
+    }
+
+    public Entity<XenoArtifactNodeComponent> CreateNode(Entity<XenoArtifactComponent> ent, ProtoId<XenoArchTriggerPrototype> trigger, int depth = 0)
+    {
+        return CreateNode(ent, PrototypeManager.Index(trigger), depth);
+    }
+
+    public Entity<XenoArtifactNodeComponent> CreateNode(Entity<XenoArtifactComponent> ent, XenoArchTriggerPrototype trigger, int depth = 0)
+    {
+        var proto = PrototypeManager.Index(ent.Comp.EffectWeights).Pick(RobustRandom);
+
+        AddNode((ent, ent), proto, out var nodeEnt, dirty: false);
```

Comment:
its actually TryAddNode. But false means it kinda crushed. Should there be some code in case it could not create?

Reply:
And this (ent, ent) is irritating...